### PR TITLE
Issue #11257 - Apply "files" npm whitelist at build time

### PR DESCRIPTION
**Pull request by yu-tian113**
Compare files/folders in package/npm to "files" entries in package.json, only copy whitelisted files/folders to build.

### CODE REVIEW

**Comment by gaearon** [comment]
I'm thinking that we also probably want to fail the build if there is a `.js` file in the package root that doesn't have an `npm/*.js` equivalent.

**Comment by yu-tian113** [comment]
Thanks for the advice @gaearon, I can do a file check after `copyBundleIntoNodePackage` resolved, and either return another promise to continue the build or fail it if files don't match.
My only question is I'm not sure in what cases, in npm consumption packages, a `.js` file can be added to package root from somewhere other than `npm/*.js`, did I miss anything or is this a safeguard for anything unexpected in the future?

**Comment by gaearon** [comment]
>I'm not sure in what cases, in npm consumption packages, a .js file can be added to package root from somewhere other than npm/*.js

We use files like `react-dom/server.js` for local testing and bundling, but we need corresponding files like `react-dom/npm/server.js` to point to compiled bundles in the built packages. We want to ensure that we can't add a new entry point like `react-dom/my-entry-point.js` that would pass the tests without also adding a corresponding `react-dom/npm/my-entry-point.js` file and adding that file to `"files"` in the `package.json`.

Does this make sense/

**Comment by yu-tian113** [comment]
Awesome @gaearon.
I see the current react-dom build will fail then because the `index.fb.js` in the package root does not exist in `react-dom/npm`. 
I'll need to create an equivalent in `react-dom/npm` to pass this file matching check. I'm thinking
```
'use strict';
module.exports = require('./index');
```

It won't be exposed in npm packages as it's not whitelisted in package.json. 
How does it sound to you?

**Comment by gaearon** [comment]
Hmm. Good point, I haven't though about this.
Let's just make an exception for `.fb.js` files?

**Comment by yu-tian113** [comment]
Hi @gaearon, please review the updated code as per our discussion above.
Lint is failing on the master branch right now, I'll follow up once it's fixed.
Cheers.

**Comment by gaearon** [comment]
Can you rebase now?

**Comment by yu-tian113** [comment]
Done, please review. Thanks @gaearon.

File: scripts/rollup/packaging.js [hunk 1]
```diff
@@ -165,15 +166,48 @@ function copyNodePackageTemplate(packageName) {
...
   }
-  // TODO: verify that all copied files are either in the "files"
-  // whitelist or implicitly published by npm.
-  return asyncCopyTo(npmFrom, to).then(() =>
-    Promise.all([
+  fs.mkdirSync(to);
+  const packageJson = resolve(`${from}/package.json`);
+  const whitelistedFiles = fs.existsSync(packageJson)
+    ? require(packageJson).files
+    : null;
```
Comment (gaearon) [inline scripts/rollup/packaging.js]: Maybe just fall back to `[]`? Then you don't have to check for existence later.

File: scripts/rollup/packaging.js [hunk 2]
```diff
@@ -165,15 +166,48 @@ function copyNodePackageTemplate(packageName) {
...
+    // Terminate the build if any entry point in the package root does not have an equivalent in ./npm.
+    if (!npmRootFiles.includes(entry)) {
+      console.error(
+        `Entry point ${entry} in package ${packageName} does not have an equivalent in ./npm`
+      );
+      process.exit(1);
+    }
+  });
+  if (whitelistedFiles && whitelistedFiles.length > 0) {
+    let asyncCopyProxies = [];
```
Comment (gaearon) [inline scripts/rollup/packaging.js]: Useless assignment: it's reassigned right afterwards. Let's combine?

File: scripts/rollup/packaging.js [hunk 3]
```diff
@@ -165,15 +166,48 @@ function copyNodePackageTemplate(packageName) {
...
+    asyncCopyProxies = npmRootFiles.reduce((proxies, file) => {
+      if (
```
Comment (gaearon) [inline scripts/rollup/packaging.js]: A comment here would help understand what this does and why

File: scripts/rollup/packaging.js [hunk 4]
```diff
@@ -165,15 +166,48 @@ function copyNodePackageTemplate(packageName) {
...
+        whitelistedFiles.includes(`${file}/`)
+      ) {
+        proxies.push(
+          asyncCopyTo(resolve(`${npmFrom}/${file}`), `${to}/${file}`)
+        );
+      }
+      return proxies;
+    }, asyncCopyProxies);
+    return Promise.all([
+      ...asyncCopyProxies,
```
Comment (gaearon) [inline scripts/rollup/packaging.js]: Maybe `promisesForForwardingModules`?

File: scripts/rollup/packaging.js (hunk 1 above)
Comment (yu-tian113) [reply scripts/rollup/packaging.js]: Agree. Also need to check `require(packageJson).files` is not `undefined`, as **files** entry is missing in some package.json(e.g. react-call-return).
Updated to 
`const whitelistedFiles = (fs.existsSync(packageJson) &&
    require(packageJson).files) || [];` 
And deleted the existance check below.
~~`if (whitelistedFiles && whitelistedFiles.length > 0) {}`~~
`if (whitelistedFiles.length > 0) {}`

File: scripts/rollup/packaging.js (hunk 2 above)
Comment (yu-tian113) [reply scripts/rollup/packaging.js]: Agree, done.

File: scripts/rollup/packaging.js (hunk 3 above)
Comment (yu-tian113) [reply scripts/rollup/packaging.js]: Comments added

File: scripts/rollup/packaging.js (hunk 4 above)
Comment (yu-tian113) [reply scripts/rollup/packaging.js]: hehe, sorry, didn't realise I typed these **promises** as **proxies**.
Variables renamed.

**Comment by yu-tian113** [comment]
Thanks @gaearon. All updated, please see my comments.

**Comment by gaearon** [comment]
Shouldn't we also fail if an entry point is not in `files` in `package.json`?

**Comment by yu-tian113** [comment]
Good point, the problem is some package like `react-call-return` omits the `files` field, which indicates 

> everything except automatically-excluded files will be included in the publish

I'm going to check the existence of `files` field, if not exists, everything in `./npm` are whitelisted and will be copied to build.

Another defect l noticed, wildcard patterns(`"glob/*.{js,json}"`) in `files` field is not supported properly in the current solution.

Working on these two issues now, talk soon.

**Comment by gaearon** [comment]
>Good point, the problem is some package like react-call-return omits the files field

This actually sounds like a problem we should fix in that package. 😉 
Let's mandate that the field exists.

**Comment by gaearon** [comment]
I fixed `react-call-return` in https://github.com/facebook/react/commit/b2ff29d38c23e32fca6e6f553ba743276bb77282. We should throw if `files` doesn't exist unless it is a `private` package (in which case it's fine).

**Comment by yu-tian113** [comment]
No problem, thanks for the fix.
Just done some updates to handle these and also the wildcard pattern issue I mentioned above.
All the checks we added should work with globs like `*.js` in `files` now.

File: scripts/rollup/packaging.js [hunk 5]
```diff
@@ -165,15 +167,81 @@ function copyNodePackageTemplate(packageName) {
...
+  const packageJson = `${from}/package.json`;
+  const whitelist = fs.existsSync(packageJson) && require(packageJson).files;
+  let promisesForForwardingModules = [];
+  if (!whitelist) {
+    // Terminate the build if 'files' field is missing from package.json.
+    console.error(
+      `'files' field is missing from package.json in package ${packageName}`
+    );
+    process.exit(1);
+  } else {
```
Comment (gaearon) [inline scripts/rollup/packaging.js]: Since this is a hard exit, can you remove the `else` and just leave it after this code unindented?

**Comment by yu-tian113** [comment]
Yes, makes sense, please review, thanks @gaearon.

**Comment by gaearon** [comment]
Seems like file permissions were changed accidentally?

```
scripts/rollup/packaging.js 100644 → 100755
```

File: scripts/rollup/packaging.js [hunk 6]
```diff
@@ -165,15 +167,80 @@ function copyNodePackageTemplate(packageName) {
...
+    // Terminate the build if 'files' field is missing from package.json.
+    console.error(
+      `'files' field is missing from package.json in package ${packageName}`
+    );
+    process.exit(1);
+  }
+  fs.mkdirSync(to);
+  let existingDirCache = {};
+  // looping through entries(single file / directory / pattern) in `files`
+  const whitelistedFiles = whitelist.reduce((list, pattern) => {
```
Comment (gaearon) [inline scripts/rollup/packaging.js]: This logic is a bit complicated. Is there any module we can use for this? Or maybe there's some simpler way to write it? At least let's extract it to a separate function.

Comment (yu-tian113) [reply scripts/rollup/packaging.js]: Thanks @gaearon, some modules may help with bits and pieces, but I'm a little hesitant including any as new dev dependencies, considering we already use things like `ncp` and `glob`(which the other modules are build upon). Also `glob` composes a whiteListedFiles list used for checking all entry points are whitelisted. If a module doing things like `copy with glob`, we'll have to run `glob` again to get the list.
[node-mkdirp](https://github.com/substack/node-mkdirp) can help creating nested directories. Do you think I can add it as dev dependency?
I will extract it to a separate function, make the code flow more intuitive.

Comment (gaearon) [reply scripts/rollup/packaging.js]: node-mkdirp seems fine. Why didn't we need it before?

**Comment by yu-tian113** [comment]
Thanks @gaearon, `mkdirp` creates parent directories on the fly, help us get rid of the **reduce inside reduce** in the old code :)
I didn't know about this module before you suggested making the code simpler. Thought I could just build my own in about 10 lines, but it did complicate the function a lot.
I also reset the file permission back to `644`.
Please review.

**Comment by gaearon** [comment]
This still seems complicated. My biggest issue is we're essentially reimplementing npm's `files` inclusion logic.

I think it would make more sense if we used `npm pack` instead of copying the directory. Then we'd literally test what `npm publish` actually does (but "deploy" to our build folder): http://podefr.tumblr.com/post/30488475488/locally-test-your-npm-modules-without-publishing

What do you think?

**Comment by yu-tian113** [comment]
Sorry, I'm a little lost here. Do you mind explaining the idea using `npm pack`?
How will it deal with the file **replacement**(entry points) and **copy**(Circle.js... in react-art package) from `/npm`?
Besides, reimplementing the `files` inclusion logic also make sure the build will warn and fail if an entry point is not whitelisted.
Thanks!

**Comment by gaearon** [comment]
I was thinking that maybe we need to change the build step to be:

* Build to a temporary folder
* npm pack it
* Unpack into our `build/packages` folder structure

This way we simulate what would actually happen during npm publish.

>Besides, reimplementing the files inclusion logic also make sure the build will warn and fail if an entry point is not whitelisted.

That's true. But I added a suite of tests that run on bundles in `build/` folder a week ago. So if something is missing there, those tests would fail. In other words, instead of trying to validate it ourselves, we can just "simulate" a publish and then run the tests on the packages folder.

**Comment by yu-tian113** [comment]
> Build to a temporary folder

Here we'll still copy **everything** inside `/npm` directory to the temporary folder as we've been doing.
> npm pack it
> Unpack into our build/packages folder structure

I think npm pack and unpack is pretty handy to **prepare** the build packages, but not much of help **validating** the packages, as they just silently skip any missing files. So all the validation need to be handled in the test suite.

I'm happy to give it a try.

**Comment by gaearon** [comment]
>So all the validation need to be handled in the test suite.

Yes, but this is enough because our test suite covers all public API (or at least it should—if not, we should fix the test suite), and >92% tests run on bundles.

**Comment by yu-tian113** [comment]
Hi @gaearon, I've created a new pull request [#11750](https://github.com/facebook/react/pull/11750) for the updates as per our discussion above.
Please review, thanks!

**Comment by gaearon** [comment]
We decided to go with https://github.com/facebook/react/pull/11750 instead.

**Pull request closed without merging**